README.md

HubSpot Connector
=================

HubSpot all-in-one marketing software helps more than 8,000 companies in 56 countries attract leads and convert them into customers. A pioneer in inbound marketing, HubSpot aims to help its customers make marketing that people actually love.

Installation and Usage
----------------------

For information about usage and installation you can check our documentation at http://mulesoft.github.com/hubspot-connector.

Reporting Issues
----------------

We use GitHub:Issues for tracking issues with this connector. You can report new issues at this link https://github.com/mulesoft/hubspot-connector/issues

Find something you don't like?
------------------------------

Want to add your own magic to the connector? Go to [Contributing to Anypoint Community Connectors](http://mulesoft.github.io/connector-certification-docs/contr/index.html) and learn how you can also be a part of it.


CHANGELOG.md
Hubspot Connector Change Log
============================

## [3.0.0] - 2015-11-26
- Added new operation getAllContactsUpdatedAfter that returns all the recent contacts filtered by a waterMark
- Migrate connector to DevKit 3.7.2.

## [2.7] - 2013-04-25
- Added new operation getRecentContactsPaginated that implements Iterable and returns an iterator that handles the Pagination on the background

## [2.6.10] - 2013-04-17
- Added new version of the ObjectStore for Integration Tests. When an object is required from the OS the result is cloned and returned a new instance. This is to simulate the stateless OS from CloudHub.

## [2.6.9] - 2013-04-08
- Minor fix

## [2.6.8] - 2013-04-08
- Minor fix in refresh operation (passed object store instead credentials object)

## [2.6.7] - 2013-04-08
-  The new credentials in the refresh are stored in the ObjectStore

## [2.6.6] - 2013-04-08
-  Added log lines in refresh
- Added synchronization to the refresh operation

## [2.6.5] - 2013-04-03
-  Added new test to check that refresh token is not called again after a successful refresh operation	
- Minor fixes IT tests

## [2.6.4] - 2013-03-26
-  Fixed javadoc for scope. Use space character instead +.
- Added server error messages when operation call fails

## [2.6.3] - 2013-03-13
-  Added refresh token logic. (Requirement: offline scope) If some of the operations throws an HubSpotConnectorAccessTokenExpiredException the connector will try to get a new token using the refresh token and the refresh token endpoint 
- Added IT for the case
- Added new parameter for OAuthCredentials - offlineScope, to know if the scope enables the refresh token functionality

## [2.6.2] - 2013-03-05
- Added new operation createContactList from List API
	
## [2.6.1] - 2013-03-04
- Added new operation addExistingContactInAList form List API
- Added integration tests for new operation

## [2.6] - 2013-03-01
- Added hubId (PortalId) in all the Email API operations as @Optional parameter
- Added new email operation updateEmailSubscriptionStatusUnsubscribeFromAll

## [2.5.1] - 2013-02-28
- Added test for case ObjectStore serialization
- Moved client out of the ObjectStore and gestionated in memory
- Added clientId to the OAuthCredentials

## [2.5] - 2013-02-27
- Added config parameters to the authenticate operation in order to add more Multi Tenancy flexibility
- Minor modifications in the Integration Tests

## [2.4] - 2013-02-19
- Renamed operation from getAllProperties to getAllCustomProperties
- Addded opertaions for all the Custom Properties API

## [2.3.2] - 2013-02-18
- Added support for ObjectStore for storing the credentials. Also a configurable param for defining the Object Store to use (objectStore)

## [2.3.1] - 2013-02-15
- Create Contact returns a Contact entity instead of ContactProperties

## [2.3] - 2013-02-14
- Added operations for all the Email API

## [2.2.2] - 2013-02-14
- Added static method inside ContactPropertiesNumberOfEmployees getValueFrom

## [2.2.1] - 2013-02-06
- Added support for Custom Properties inside Contact. Added Map customProperties.

## [2.2] - 2013-02-06
- Added new Operation for the Contact Properties API: getAllProperties
- Added unit and integration tests

## [2.1] - 2013-02-06
- Added more unit and integration tests
-Changed log level from info to debug (require extra parameters in log4j)
- Changed signature in operations of the Lists API from Strings (JSON Format) to POJOs

## [2.0] - 2013-02-06
- Added integration test for all the Contacts API
- Changed signature in operations of the Contacts API from Strings (JSON Format) to POJOs

## [1.2.2] - 2013-02-06
- Added structure for Integration Tests

## [1.2.1] - 2013-02-05
- Added Unit Test for POJOs transformations

## [1.2] - 2013-02-05
- Created POJOs structure. Does not affect Connector Process signatures neither configuration
- Added custom Serializer/Deserializer for ContactProperties

## [1.1] - 2013-02-05
- Mass refactor. Does not affect Connector Process signatures neither configuration

## [1.0.1] - 2013-02-05
- Added new process Get Contacts in a List
	
## [1.0] - 2013-12-07
- First Release. Alpha version
- Implementation of the Contact API
- OAuth2 implementation through authenticate and authenticateResponse methods (see demo for a example flow)
- Exceptions hierarchy for a easy comprehension of the problems that may occur in the connector
- Responses of the process are in JSON format



java/lists.js
var DATA = [
      { id:0, label:"authenticate", link:"mule/hubspot-config.html#authenticate", type:"method" },
      { id:1, label:"authenticate-response", link:"mule/hubspot-config.html#authenticate-response", type:"method" },
      { id:2, label:"create-contact", link:"mule/hubspot-config.html#create-contact", type:"method" },
      { id:3, label:"delete-contact", link:"mule/hubspot-config.html#delete-contact", type:"method" },
      { id:4, label:"get-all-contacts", link:"mule/hubspot-config.html#get-all-contacts", type:"method" },
      { id:5, label:"get-contact-by-email", link:"mule/hubspot-config.html#get-contact-by-email", type:"method" },
      { id:6, label:"get-contact-by-id", link:"mule/hubspot-config.html#get-contact-by-id", type:"method" },
      { id:7, label:"get-contact-by-user-token", link:"mule/hubspot-config.html#get-contact-by-user-token", type:"method" },
      { id:8, label:"get-contact-list-by-id", link:"mule/hubspot-config.html#get-contact-list-by-id", type:"method" },
      { id:9, label:"get-contact-statistics", link:"mule/hubspot-config.html#get-contact-statistics", type:"method" },
      { id:10, label:"get-contacts-by-query", link:"mule/hubspot-config.html#get-contacts-by-query", type:"method" },
      { id:11, label:"get-contacts-lists", link:"mule/hubspot-config.html#get-contacts-lists", type:"method" },
      { id:12, label:"get-dynamic-contact-lists", link:"mule/hubspot-config.html#get-dynamic-contact-lists", type:"method" },
      { id:13, label:"get-recent-contacts", link:"mule/hubspot-config.html#get-recent-contacts", type:"method" },
      { id:14, label:"has-user-access-token", link:"mule/hubspot-config.html#has-user-access-token", type:"method" },
      { id:15, label:"org.mule.module.hubspot", link:"java/org/mule/module/hubspot/package-summary.html", type:"package" },
      { id:16, label:"org.mule.module.hubspot.HubSpotConceptProof", link:"java/org/mule/module/hubspot/HubSpotConceptProof.html", type:"class" },
      { id:17, label:"org.mule.module.hubspot.HubSpotConnector", link:"java/org/mule/module/hubspot/HubSpotConnector.html", type:"class" },
      { id:18, label:"org.mule.module.hubspot.adapters", link:"java/org/mule/module/hubspot/adapters/package-summary.html", type:"package" },
      { id:19, label:"org.mule.module.hubspot.adapters.HubSpotConnectorCapabilitiesAdapter", link:"java/org/mule/module/hubspot/adapters/HubSpotConnectorCapabilitiesAdapter.html", type:"class" },
      { id:20, label:"org.mule.module.hubspot.adapters.HubSpotConnectorLifecycleAdapter", link:"java/org/mule/module/hubspot/adapters/HubSpotConnectorLifecycleAdapter.html", type:"class" },
      { id:21, label:"org.mule.module.hubspot.adapters.HubSpotConnectorMetadataAdapater", link:"java/org/mule/module/hubspot/adapters/HubSpotConnectorMetadataAdapater.html", type:"class" },
      { id:22, label:"org.mule.module.hubspot.adapters.HubSpotConnectorProcessAdapter", link:"java/org/mule/module/hubspot/adapters/HubSpotConnectorProcessAdapter.html", type:"class" },
      { id:23, label:"org.mule.module.hubspot.agents", link:"java/org/mule/module/hubspot/agents/package-summary.html", type:"package" },
      { id:24, label:"org.mule.module.hubspot.agents.DefaultSplashScreenAgent", link:"java/org/mule/module/hubspot/agents/DefaultSplashScreenAgent.html", type:"class" },
      { id:25, label:"org.mule.module.hubspot.config", link:"java/org/mule/module/hubspot/config/package-summary.html", type:"package" },
      { id:26, label:"org.mule.module.hubspot.config.AbstractDefinitionParser", link:"java/org/mule/module/hubspot/config/AbstractDefinitionParser.html", type:"class" },
      { id:27, label:"org.mule.module.hubspot.config.AbstractDefinitionParser.ParseDelegate", link:"java/org/mule/module/hubspot/config/AbstractDefinitionParser.ParseDelegate.html", type:"class" },
      { id:28, label:"org.mule.module.hubspot.config.AuthenticateDefinitionParser", link:"java/org/mule/module/hubspot/config/AuthenticateDefinitionParser.html", type:"class" },
      { id:29, label:"org.mule.module.hubspot.config.AuthenticateResponseDefinitionParser", link:"java/org/mule/module/hubspot/config/AuthenticateResponseDefinitionParser.html", type:"class" },
      { id:30, label:"org.mule.module.hubspot.config.CreateContactDefinitionParser", link:"java/org/mule/module/hubspot/config/CreateContactDefinitionParser.html", type:"class" },
      { id:31, label:"org.mule.module.hubspot.config.DeleteContactDefinitionParser", link:"java/org/mule/module/hubspot/config/DeleteContactDefinitionParser.html", type:"class" },
      { id:32, label:"org.mule.module.hubspot.config.GetAllContactsDefinitionParser", link:"java/org/mule/module/hubspot/config/GetAllContactsDefinitionParser.html", type:"class" },
      { id:33, label:"org.mule.module.hubspot.config.GetContactByEmailDefinitionParser", link:"java/org/mule/module/hubspot/config/GetContactByEmailDefinitionParser.html", type:"class" },
      { id:34, label:"org.mule.module.hubspot.config.GetContactByIdDefinitionParser", link:"java/org/mule/module/hubspot/config/GetContactByIdDefinitionParser.html", type:"class" },
      { id:35, label:"org.mule.module.hubspot.config.GetContactByUserTokenDefinitionParser", link:"java/org/mule/module/hubspot/config/GetContactByUserTokenDefinitionParser.html", type:"class" },
      { id:36, label:"org.mule.module.hubspot.config.GetContactListByIdDefinitionParser", link:"java/org/mule/module/hubspot/config/GetContactListByIdDefinitionParser.html", type:"class" },
      { id:37, label:"org.mule.module.hubspot.config.GetContactStatisticsDefinitionParser", link:"java/org/mule/module/hubspot/config/GetContactStatisticsDefinitionParser.html", type:"class" },
      { id:38, label:"org.mule.module.hubspot.config.GetContactsByQueryDefinitionParser", link:"java/org/mule/module/hubspot/config/GetContactsByQueryDefinitionParser.html", type:"class" },
      { id:39, label:"org.mule.module.hubspot.config.GetContactsListsDefinitionParser", link:"java/org/mule/module/hubspot/config/GetContactsListsDefinitionParser.html", type:"class" },
      { id:40, label:"org.mule.module.hubspot.config.GetDynamicContactListsDefinitionParser", link:"java/org/mule/module/hubspot/config/GetDynamicContactListsDefinitionParser.html", type:"class" },
      { id:41, label:"org.mule.module.hubspot.config.GetRecentContactsDefinitionParser", link:"java/org/mule/module/hubspot/config/GetRecentContactsDefinitionParser.html", type:"class" },
      { id:42, label:"org.mule.module.hubspot.config.HasUserAccessTokenDefinitionParser", link:"java/org/mule/module/hubspot/config/HasUserAccessTokenDefinitionParser.html", type:"class" },
      { id:43, label:"org.mule.module.hubspot.config.HubSpotConnectorConfigDefinitionParser", link:"java/org/mule/module/hubspot/config/HubSpotConnectorConfigDefinitionParser.html", type:"class" },
      { id:44, label:"org.mule.module.hubspot.config.HubspotNamespaceHandler", link:"java/org/mule/module/hubspot/config/HubspotNamespaceHandler.html", type:"class" },
      { id:45, label:"org.mule.module.hubspot.config.UpdateContactDefinitionParser", link:"java/org/mule/module/hubspot/config/UpdateContactDefinitionParser.html", type:"class" },
      { id:46, label:"org.mule.module.hubspot.exception", link:"java/org/mule/module/hubspot/exception/package-summary.html", type:"package" },
      { id:47, label:"org.mule.module.hubspot.exception.HubSpotConnectorAccessTokenExpiredException", link:"java/org/mule/module/hubspot/exception/HubSpotConnectorAccessTokenExpiredException.html", type:"class" },
      { id:48, label:"org.mule.module.hubspot.exception.HubSpotConnectorException", link:"java/org/mule/module/hubspot/exception/HubSpotConnectorException.html", type:"class" },
      { id:49, label:"org.mule.module.hubspot.exception.HubSpotConnectorNoAccessTokenException", link:"java/org/mule/module/hubspot/exception/HubSpotConnectorNoAccessTokenException.html", type:"class" },
      { id:50, label:"org.mule.module.hubspot.model", link:"java/org/mule/module/hubspot/model/package-summary.html", type:"package" },
      { id:51, label:"org.mule.module.hubspot.model.OAuthCredentials", link:"java/org/mule/module/hubspot/model/OAuthCredentials.html", type:"class" },
      { id:52, label:"org.mule.module.hubspot.process", link:"java/org/mule/module/hubspot/process/package-summary.html", type:"package" },
      { id:53, label:"org.mule.module.hubspot.process.NestedProcessorChain", link:"java/org/mule/module/hubspot/process/NestedProcessorChain.html", type:"class" },
      { id:54, label:"org.mule.module.hubspot.process.ProcessCallbackProcessInterceptor", link:"java/org/mule/module/hubspot/process/ProcessCallbackProcessInterceptor.html", type:"class" },
      { id:55, label:"org.mule.module.hubspot.processors", link:"java/org/mule/module/hubspot/processors/package-summary.html", type:"package" },
      { id:56, label:"org.mule.module.hubspot.processors.AbstractExpressionEvaluator", link:"java/org/mule/module/hubspot/processors/AbstractExpressionEvaluator.html", type:"class" },
      { id:57, label:"org.mule.module.hubspot.processors.AbstractListeningMessageProcessor", link:"java/org/mule/module/hubspot/processors/AbstractListeningMessageProcessor.html", type:"class" },
      { id:58, label:"org.mule.module.hubspot.processors.AbstractMessageProcessor", link:"java/org/mule/module/hubspot/processors/AbstractMessageProcessor.html", type:"class" },
      { id:59, label:"org.mule.module.hubspot.processors.AuthenticateMessageProcessor", link:"java/org/mule/module/hubspot/processors/AuthenticateMessageProcessor.html", type:"class" },
      { id:60, label:"org.mule.module.hubspot.processors.AuthenticateResponseMessageProcessor", link:"java/org/mule/module/hubspot/processors/AuthenticateResponseMessageProcessor.html", type:"class" },
      { id:61, label:"org.mule.module.hubspot.processors.CreateContactMessageProcessor", link:"java/org/mule/module/hubspot/processors/CreateContactMessageProcessor.html", type:"class" },
      { id:62, label:"org.mule.module.hubspot.processors.DeleteContactMessageProcessor", link:"java/org/mule/module/hubspot/processors/DeleteContactMessageProcessor.html", type:"class" },
      { id:63, label:"org.mule.module.hubspot.processors.GetAllContactsMessageProcessor", link:"java/org/mule/module/hubspot/processors/GetAllContactsMessageProcessor.html", type:"class" },
      { id:64, label:"org.mule.module.hubspot.processors.GetContactByEmailMessageProcessor", link:"java/org/mule/module/hubspot/processors/GetContactByEmailMessageProcessor.html", type:"class" },
      { id:65, label:"org.mule.module.hubspot.processors.GetContactByIdMessageProcessor", link:"java/org/mule/module/hubspot/processors/GetContactByIdMessageProcessor.html", type:"class" },
      { id:66, label:"org.mule.module.hubspot.processors.GetContactByUserTokenMessageProcessor", link:"java/org/mule/module/hubspot/processors/GetContactByUserTokenMessageProcessor.html", type:"class" },
      { id:67, label:"org.mule.module.hubspot.processors.GetContactListByIdMessageProcessor", link:"java/org/mule/module/hubspot/processors/GetContactListByIdMessageProcessor.html", type:"class" },
      { id:68, label:"org.mule.module.hubspot.processors.GetContactStatisticsMessageProcessor", link:"java/org/mule/module/hubspot/processors/GetContactStatisticsMessageProcessor.html", type:"class" },
      { id:69, label:"org.mule.module.hubspot.processors.GetContactsByQueryMessageProcessor", link:"java/org/mule/module/hubspot/processors/GetContactsByQueryMessageProcessor.html", type:"class" },
      { id:70, label:"org.mule.module.hubspot.processors.GetContactsListsMessageProcessor", link:"java/org/mule/module/hubspot/processors/GetContactsListsMessageProcessor.html", type:"class" },
      { id:71, label:"org.mule.module.hubspot.processors.GetDynamicContactListsMessageProcessor", link:"java/org/mule/module/hubspot/processors/GetDynamicContactListsMessageProcessor.html", type:"class" },
      { id:72, label:"org.mule.module.hubspot.processors.GetRecentContactsMessageProcessor", link:"java/org/mule/module/hubspot/processors/GetRecentContactsMessageProcessor.html", type:"class" },
      { id:73, label:"org.mule.module.hubspot.processors.HasUserAccessTokenMessageProcessor", link:"java/org/mule/module/hubspot/processors/HasUserAccessTokenMessageProcessor.html", type:"class" },
      { id:74, label:"org.mule.module.hubspot.processors.UpdateContactMessageProcessor", link:"java/org/mule/module/hubspot/processors/UpdateContactMessageProcessor.html", type:"class" },
      { id:75, label:"update-contact", link:"mule/hubspot-config.html#update-contact", type:"method" }

    ];
